File: 백준/Gold/21943. 연산 최대로/연산 최대로.js
const fs = require("fs");
const [L1, L2, L3] = fs
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const N = +L1;
const nums = L2.split(" ").map(Number);
let ops = L3.split(" ").map(Number);
const P = ops[0];
const Q = ops[1];
let ans = 0;
function permutaion(depth, sel, v) {
  if (depth === N + P + Q) {
    let tmp = sel[0];
    let total = 1;
    for (let i = 1; i < sel.length - 1; i += 2) {
      let num = sel[i + 1];
      let op = sel[i];
      if (op === 1) {
        total *= tmp;
        tmp = num;
      } else {
        tmp += num;
      }
    }
    total *= tmp;
    ans = Math.max(ans, total);
    return;
  }
  if (depth % 2 === 0) {
    for (let i = 0; i < N; i++) {
      if (!v[i]) {
        v[i] = true;
        sel.push(nums[i]);
        permutaion(depth + 1, sel, v);
        sel.pop();
        v[i] = false;
      }
    }
  } else {
    for (let i = 0; i < 2; i++) {
      if (ops[i] !== 0) {
        ops[i]--;
        sel.push(i);
        permutaion(depth + 1, sel, v);
        sel.pop();
        ops[i]++;
      }
    }
  }
}

permutaion(0, [], new Array(N).fill(false));

console.log(ans);

File: 백준/Gold/21943. 연산 최대로/연산 최대로.java
import java.util.*;
import java.io.*;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

	static void permutaion(int depth, boolean[] v, int[] select) {
		if (depth == N + P + Q) {
			long total = select[0];
			long tmp = 1;
			for(int i = 1 ; i < select.length - 1 ; i += 2) {
				int op = select[i];
				long num = select[i + 1];
				if(op == 0) {
					total += num;
				}else {
					tmp *= total;
					total = num;
				}
			}
			tmp *= total;
			ans = Math.max(ans, tmp);
			return;
		}
		if(depth % 2 == 0) {
			for(int i = 0 ; i < N ; i++) {
				if(!v[i]) {
					v[i] = true;
					select[depth] = nums[i]; 
					permutaion(depth + 1, v, select);
					v[i] = false;
				}
			}
		}else {
			for(int i = 0 ; i < 2 ; i++) {
				if(ops[i] != 0) {
					ops[i]--;
					select[depth] = i;
					permutaion(depth + 1, v, select);
					ops[i]++;
				}
			}
		}

	}

	static long ans = 0;
	static int N, P, Q = 0;
	static int[] nums;
	static int[] ops;
	
	public static void main(String[] args) throws IOException {
		N = Integer.parseInt(br.readLine());
		StringTokenizer st = new StringTokenizer(br.readLine());
		nums = new int[N];
		for (int i = 0; i < N; i++) {
			nums[i] = Integer.parseInt(st.nextToken());
		}
		st = new StringTokenizer(br.readLine());
		ops = new int[2];
		P = ops[0] = Integer.parseInt(st.nextToken());
		Q = ops[1] = Integer.parseInt(st.nextToken());
		boolean[] v = new boolean[N];
		int[] select = new int[N + P + Q];
		permutaion(0, v, select);
		bw.write(ans + "");
		bw.close();

	}
}